Решение

7eAli opened the pull request:
Решение

7eAli committed:
initial commit

7eAli committed:
Первый вариант решения задачи

Liberate520 commented on src/Family_tree/Human.java:
```diff
@@ -0,0 +1,206 @@
+package Family_tree;
```
название пакетов с маленькой буквы

Liberate520 commented on src/Family_tree/Human.java:
```diff
@@ -0,0 +1,206 @@
+    private int id;
+    private String firstName, secondName, fathersName;
+    private LocalDate dateOfBirth, dateOfDeath;
+    private boolean alive = true;
+    private Gender gender;
+    private HashMap<Human,Parent_Type> parents;
+    private boolean hasBioMother = false;
+    private boolean hasBioFather = false;
```
инициализацию лучше делать в конструкторе класса

Liberate520 commented on src/Family_tree/Human.java:
```diff
@@ -0,0 +1,206 @@
+        this.dateOfBirth = dateOfBirth;
+        this.dateOfDeath = dateOfDeath;
+        this.gender = gender;
+        this.parents = new HashMap<>();
```
почему родителей инициализируете здесь, а список детей не здесь? Инициализация полей класса должна происходить в конструкторе

Liberate520 commented on src/Family_tree/Human.java:
```diff
@@ -0,0 +1,206 @@
+    }
+
+    public void addParent(Human parent, Parent_Type parentType){
+        if(this.parents.equals(null)){
+            this.parents = new HashMap<>();
+        }
```
такие конструкции бы не понадобились, если конструктор класса был описан верно

Liberate520 commented on src/Family_tree/Human.java:
```diff
@@ -0,0 +1,206 @@
+        }
+    }
+
+    public void getMarried(Human spouse){
+        this.spouse = spouse;
+        this.married = true;
+    }
```
это скорее сеттер

Liberate520 commented on src/Family_tree/Human.java:
```diff
@@ -0,0 +1,206 @@
+        }
+    }
+
+    public void getDivorced(Human spouse){
+        this.spouse = null;
+        this.married = false;
+    }
```
get это получение значения, а ваш метод ничего не возвращает

Liberate520 commented on src/Family_tree/Human.java:
```diff
@@ -0,0 +1,206 @@
+        return infoFormat();
+    }
+
+    public String infoFormat(){
```
слишком длинный метод. Должет быть разбит на более мелкие

Liberate520 commented on src/Family_tree/Main.java:
```diff
@@ -0,0 +1,43 @@
+        family_1.addFamilyMember(petya_2);
+        family_1.addFamilyMember(dasha_2);
+        family_1.addFamilyMember(oleg_2);
+        vanya_1.addParent(petya_2, Parent_Type.Biological_Father);
+        vanya_1.addParent(dasha_2, Parent_Type.Biological_Mother);
+        vanya_1.addParent(oleg_2, Parent_Type.Foster_Father);
+        vanya_1.getMarried(olya_1);
+        olya_1.setSecondName("Каруселева");
+        vanya_1.addChild(kostya_0, Child_type.Biological_Son);
+        vanya_1.addChild(petya_0, Child_type.Foster_Son);
+        vanya_1.addChild(masha_0, Child_type.Biological_Daughter);
+        vanya_1.addChild(katya_0, Child_type.Foster_Daughter);
+        olya_1.addChild(kostya_0, Child_type.Biological_Son);
+        olya_1.addChild(petya_0, Child_type.Foster_Son);
+        olya_1.addChild(masha_0, Child_type.Biological_Daughter);
+        olya_1.addChild(katya_0, Child_type.Foster_Daughter);
```
нет автоматизации. Например сейчас вы установили связи вани со многими людьми, но эти люди ничего не знают про ваню

7eAli committed:
Исправления по комментариям

7eAli commented:
Вроде бы все исправления по вашим комментариям сделал. Добавил метод свадьбы и развода в класс дерева. По поводу автоматизации - теперь когда мы добавляем родителя ребенку, то ребенок добавляется родителю и наоборот. Авто-добавление детей к супругу пока не стал делать, потому что слишком много разных ситуаций можно придумать

7eAli committed:
Второе задание

7eAli committed:
Третье задание

7eAli committed:
Четвертое задание

7eAli committed:
Пятое задание

Liberate520 commented on src/family_tree/model/tree/Tree.java:
```diff
@@ -0,0 +1,79 @@
+
+    public void wedding(E husband, E wife){
+        if(husband.isMarried() || wife.isMarried()){
+            System.out.println("Свадьба невозможна");
```
Работа с консолью должна выполняться во View

Liberate520 commented on src/family_tree/model/tree/TreeService.java:
```diff
@@ -0,0 +1,62 @@
+        family.sortByName();
+    }
+    public void getFamilyMemberInfo(int id){
+        System.out.println(family.getFamilyMember(id).toString());
```
Работа с консолью должна выполняться во View

7eAli committed:
Исправления к пятому заданию по комментариям

7eAli committed:
Задание 6
В соответствии с первым принципом убрал из классов Human и Tree некоторые методы и создал 2 класса, реализующие их.
В соответствии с четвертым принципом убрал из интерфейса Entity часть методов и сделал класс-наследник, реализующий их

Liberate521 commented on src/family_tree/model/tree/TreeService.java:
```diff
@@ -0,0 +1,72 @@
+
+public class TreeService {
+    private Tree<Human> family;
+    private FileHandler fileHandler;
```
это нарушение 5 принципа. Связь должна быть построена через интерфейс

7eAli committed:
Исправления по комментариям к 6-му ДЗ

Liberate520 commented on src/family_tree/presenter/Presenter.java:
```diff
@@ -0,0 +1,63 @@
+
+    public Presenter(View view){
+        this.view = view;
+        service = new TreeService(new Tree<Human>(), new ObjectFileHandler(), new Connection_Manager<Human>(), new Connection_Info<Human>());
```
3 и 4 аргумент сервис мог и сам создать в своем конструкторе
